feat(perf-views): Add the trends endpoint

wmak opened the pull request:
- Adds new `_range` functions for the trends endpoint
  - These 3 functions will be the defaults for performance trends, and the
  intent is to use them to calculate the aggregates of the time period
  in two buckets, the first half + second half. Then divide the two to
  see what the percentage change was to determine whether there's a
  trend or not.
- Adds the new events-trends endpoint that returns both the event data and the stats data, done this way to keep the number of queries low for trends
  - Accepts 3 possible trendFunctions p50, avg and user_misery

wmak committed:
feat(perf-views): Adding the aggregate functions and divide

- These 3 functions will be the defaults for performance trends, and the
  intent is to use them to calculate the aggregates of the time period
  in two buckets, the first half + second half. Then divide the two to
  see what the percentage change was to determine whether there's a
  trend or not.

wmak committed:
fix(discover): Remove duplicate aggregate

wmak committed:
feat: Adding the events-trends endpoint

- Accepts 3 possible trendFunctions p50, avg and user_misery
- returns both the top 5 events and their stats charts

wmak committed:
updating function+alias to consistent formatting

wmak committed:
Merge branch 'master' into feat/perf-trend-endpoint

wmak committed:
fix: Some issues from testing

- inf/nan need to be handled since python json allows them but not js
- count ranges were wrong
- fixing a bug with adding project transform

wmak committed:
fix: top events was only returning count

wmak committed:
fix: extra code for inf/nan, top events was still being very jank

wmak committed:
fix:ref: Adding tests

- user_misery wasn't actually working properly with the way the and was
  written, only the first two conditions were being used which meant
  that it essentially was a countIf since the duration condition was
  last

wmak commented on src/sentry/api/bases/organization_events.py:
```diff
@@ -156,6 +156,8 @@ def handle_results_with_meta(self, request, organization, project_ids, results):
                 if key not in meta:
                     meta[key] = "string"
+            # Remove any potential NaN or Inf
+            data = json.loads(json.dumps(data, ignore_nan=True))
```
I think alternatively I could iterate through the results here instead, but this seemed much simpler

dashed replied:
The cost is no different than iteratively traversing the `data` object, and replacing NaN and Inf with None.

wmak replied:
Nah... this is much more expensive than iterating:
https://gist.github.com/wmak/ee299f095f4a3fa9ffacd3a76a6d56cf

Have updated to iterate instead and added an optional param if we decide to include more divides in the future.

wmak commented on src/sentry/api/event_search.py:
```diff
@@ -1431,6 +1442,78 @@ def has_default(self, params):
+        ],
+        "calculated_args": [{"name": "tolerated", "fn": lambda args: args["satisfaction"] * 4.0}],
+        "aggregate": [
+            u"uniqIf(user,and(greater(duration,{tolerated:g}),and(greaterOrEquals(timestamp,toDateTime('{start}')),less(timestamp,toDateTime('{end}')))))",
```
Interesting one here, this had to be
```
and(duration>tolerated,
    and(
         timestamp > start,
         timestamp < end
     )
)
```
Because it looks like `and` only supports two terms.

evanh replied:
Clickhouse supports any number of and terms, but snuba expects everything to be binary. I'm not sure 100% why this wouldn't work for you, since this should be getting passed straight to clickhouse.

wmak committed:
ref: some missed tests from updating the range syntax

markstory commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -0,0 +1,121 @@
+        trend_column = self.trend_columns.get(function)
+        if trend_column is None:
+            return Response(
+                {"detail": "{} is not a supported trend function".format(trend_function)},
```
This could be a UnicodeError if the user provides non-ascii bytes.

markstory commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -0,0 +1,121 @@
+                    "minus({alias}2,{alias}1)".format(alias=trend_column["alias"]),
+                    "count_range({start},{end},1)".format(start=start, end=middle),
+                    "count_range({start},{end},2)".format(start=middle, end=end),
+                    "divide(count_range_2,count_range_1)",
```
Does this use aliases that are elsewhere in the query?

wmak replied:
Ah yes, I should make this clearer :+1:
the `count_range({start},{end},1)` gets aliased to `count_range_1`

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -1205,6 +1205,17 @@ def normalize(self, value):
+        try:
+            datetime.strptime(value, self.date_format)
+        except ValueError:
+            raise InvalidFunctionArgument("{} is not a date in the valid format".format(value))
```
This could create a UnicodeError if the user provides non-ascii bytes.

wmak committed:
ref: Addressing PR comments

- Moving from simplejson dump/load to iterating through the results once
  instead.
- Adding count_range to the trend_columns so its more consistent with
  the other trend columns
- Updating non unicode errors

evanh commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -0,0 +1,123 @@
+                raise ParseError(detail="You cannot view events from multiple projects.")
+
+            middle = params["start"] + timedelta(
+                seconds=(params["end"] - params["start"]).total_seconds()
```
I think you need to validate intervalRatio here. What if I pass `intervalRatio=0` or `intervalRatio=-1`?

evanh replied:
Also none of your tests pass in `intervalRatio`, is this parameter used?

wmak replied:
Ah not yet, thanks for catching this. Was from the even more experimental version of this endpoint. I'll remove it from this PR :+1:

evanh commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -0,0 +1,123 @@
+                + [
+                    trend_column["format"].format(*columns, start=start, end=middle, index="1"),
+                    trend_column["format"].format(*columns, start=middle, end=end, index="2"),
+                    "divide({alias}2,{alias}1)".format(alias=trend_column["alias"]),
```
Two things:
It would be ideal if these were passed using the function syntax in snuba if possible.
```
["divide", ["{alias}2".format(...), "{alias}1".format(...)]]
```

Also if you do that, you can pull the formatting of the aliases out and do them once.
```
tr_alias1 = "{alias}1".format(...)
tr_alias2 = "{alias}2".format(...)
...
["divide", [tr_alias2, alias1]],
["minus", [tr_alias2, alias1]],
...
```

wmak replied:
Moved to syntax format, but didn't format the aliases separately since query's `selected_columns` can only accept strings

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -1431,6 +1442,78 @@ def has_default(self, params):
+        ],
+        "result_type": "integer",
+    },
+    "divide": {
```
These functions already exist in clickhouse, why are you defining them again?

wmak replied:
Shouldn't I still define them here to handle accepting the args + result type?

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -1431,6 +1442,78 @@ def has_default(self, params):
+            DateArg("end"),
+            NumberRange("index", 1, None),
+        ],
+        "aggregate": [
```
I am very confused about how this works. The `timestamp` column doesn't exist in the transaction table, and you don't actually pass `timestamp` in the groupby or or selected columns anywhere.

wmak replied:
So afaict the `timestamp` works since snuba aliases `finish_ts as timestamp` 
eg. locally this query:
```
{
    'consistent': False, 'having': [], 'aggregations': [[u'quantile(0.95)', 'duration', u'p95']], 'dataset': 'discover', 'project': [2L, 3L], 'from_date': '2020-08-19T02:02:02', 'limit': 51, 'to_date':'2020-08-20T02:02:02', 'offset': 0, 'conditions': [['type', '=', u'transaction'], ('project_id', 'IN', [3L, 2L])],
    'selected_columns': ['timestamp'], 'groupby': ['timestamp']
}
```
generates this sql:
```
SELECT
    finish_ts AS timestamp,
    quantile(0.95)(duration) AS p95,
    timestamp
FROM transactions_local
WHERE (finish_ts >= toDateTime('2020-08-19T01:58:02', 'Universal')) AND (finish_ts < toDateTime('2020-08-20T01:58:02', 'Universal')) AND (project_id
IN (2, 3)) AND (('transaction' AS type) = 'transaction') AND (project_id IN (3, 2))
GROUP BY timestamp
LIMIT 0, 51
```

Zylphrex commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -0,0 +1,123 @@
+
+            middle = params["start"] + timedelta(
+                seconds=(params["end"] - params["start"]).total_seconds()
+                / (1 / float(request.GET.get("intervalRatio", 0.5)))
```
Why divide by `1/x` when you can multiply by `x`?

wmak replied:
:+1: Removed for this iteration of the function, but good point

wmak committed:
ref: Addressing PR comments

- Removing the (currently) unused intervalRatio parameter
- Updating some errors
  - Moving from Response to ParseError :bike:
  - Including an example of a valid date format
- Passing divide/minus in the correct snuba syntax
  - Had to move generation of aggregate `derived_columns` into its own
    "loop", since divide/minus take existing function aliases as
    arguments
- Noticed a lack of test for minus

wmak commented on src/sentry/api/event_search.py:
```diff
@@ -1205,6 +1205,17 @@ def normalize(self, value):
+        try:
+            datetime.strptime(value, self.date_format)
+        except ValueError:
+            raise InvalidFunctionArgument(u"{} is not a date in the valid format".format(value))
```
:+1: good idea, I put an example here, since I figure that's more human readable than the format string

wmak committed:
Merge branch 'master' into feat/perf-trend-endpoint
